### ISSUE TITLE
Major Issue: Nodes with multiple inputs/outputs cannot be represented in the current format

### ISSUE THREAD

**Issue by sheiksadique** [neuromorphs/NIR#28]
The current graph representation assumes a single input/output for a node.

It doesn't have any mechanism to work or address specific inputs/outputs of a given module. This means the input/output nodes of a graph/subgraph have an unclear role/usage. This also renders subgraphs to be of very limited use.

There are two possible solutions to this:

1. Each node has a list of inputs and outputs. And for each output node have a list of edges that lead to specific input of a subsequent input node. Ie. the graph edge would look something like `NodeA.O[2] -> NodeB.I[0]`.
2. A second option would be to adopt ONNX/function like representation. `NodeB(NodeA.O[2], ... ) -> (NodeB.O[0], NodeB.O[1])`

Are there any proposals on how to deal with this?

**Comment by matjobst** [comment]
I also like the subscript approach, but since we reference the nodes with their name when defining the edges, I would propose a fully string-based version that looks more object-oriented, so one edge would be for example ("in", "nodeB.in1"). That would signify an edge from the toplevel graph input "in" to the input "in1" of the subgraph "nodeB".

This is an example of how this could look like in practice: 
```python
 inner_model = nir.NIRGraph(
    nodes={
        "in1": nir.Input(np.array([4, 5, 2])),
        "in2": nir.Input(np.array([4, 5, 2])),
        "flat": nir.Flatten(0),
        "out1": nir.Output(np.array([20, 2])),
        "out2": nir.Output(np.array([20, 2])),
    },
    edges=[
        ("in1", "flat"),
        ("in2", "flat"),
        ("flat", "out1"),
        ("flat", "out2")
    ]
)
outer_graph = nir.NIRGraph(
    nodes = {
        "in": nir.Input(np.array([4,5,2])),
        "inner": inner_graph,
        "out": nir.Output(np.array([20,2]))
    },
    edges=[
        ("in", "inner.in1"),
        ("in", "inner.in2"),
        ("inner.out1", "out"),
        ("inner.out2", "out")
    ]
)
```

**Comment by sheiksadique** [comment]
Thanks for the quick PR @matjobst . I have to highlight that this issue is not limited to Graphs but also Nodes because there could very well be nodes/modules that take in multiple inputs that have different meaning and then produce multiple values as outputs.

So the fix need to be extended to the Node definition perhaps and not just to the Graph

**Comment by DylanMuir** [comment]
Hi all, I hope it's ok that I jump in!

The graph representation in Rockpool addresses these issues, and so could serve as an example of one possible approach to solve the problem (in a battle-tested form).

You can get the overview from the Rockpool docs here: https://rockpool.ai/advanced/graph_overview.html

Basically, we have `GraphModule` objects (equivalent to NIR `Node`s, I believe), in which we enumerate the input and output channels individually. `GraphModule`s are connected together using "graph edge" objects (called `GraphNode`s in Rockpool), which represent connectivity only and no computation.

The result is we have full flexibility on connectivity, including splits, joins, skips, partial connectivity, etc. We always use the graph edge objects between `GraphModule`s, so all the graph traversal machinery knows how to handle them.

If you would like me to clarify any of the design on a call, I'm very happy to do so.

**Comment by Jegp** [comment]
Thanks for the explanations. I think it made it clearer to me, although I'm still not convinced of a good way to go about this. I think it'd be tremendously helpful if you could tell us your thinking about the Rockpool graphs @DylanMuir! And please, by all means jump in as often as you can!

Would it be helpful to schedule a call for those interested to (a) fully map the problem, and (b) discuss the Rockpool graphs as a solution? If so, throw a :+1: and I'll send out a Doodle.

**Comment by DylanMuir** [comment]
Sure Jens, feel free to send a doodle.

**Comment by matjobst** [comment]
Here are my thoughts about the current proposals:

#### subscript string-based version (á la `edges=[("in", "inner.in1")]`):
- positive:
  - this does not require any code changes in NIR
  - in my opinion this is human readable and seems intuitive to me (you plug into named inputs and outputs)
  - this is not even limited to subgraphs, we can also easily define nodes to have multiple inputs and outputs. This would only be a matter of documentation (and then of course correct implementation in the target frameworks).
- negative:
  - this affects graph traversal code (but that would likely be true for any solution)
  - string parsing required when parsing the graph
  - good documentation on inputs and outputs of multi-in and multi-out nodes required
  - not "safe", you can easily create invalid graphs, but that was already true before because of the string-name based edges
  - open point: how do we attach to nodes with single inputs and outputs: do we still require a name for the inputs and outputs? This could be a source of unneded verbosity. If we do not require a name, then there might be ambiguity.
  
#### @sheiksadique 's `Project` proposal:
- positive:
  - nodes stay "simple", they always have one input and output
  - no string parsing for graph traversal required
- negative:
  - requires an additional node type
  - requires additional nodes in the graph
  - in my opinion less human readable and slightly unintuitive
  - you can still create invalid graphs, but push one possibly "unsafe" point from the edge definition into the `Project` node
  - good documentation on inputs and outputs of multi-in and multi-out nodes required

**Comment by Jegp** [comment]
Closed with #50
